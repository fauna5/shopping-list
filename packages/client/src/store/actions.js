import axios from 'axios';

import { UPDATE_LIST } from './mutations';

export const GET_LIST = 'GET_LIST';
export const ADD = 'ADD';
export const UPDATE = 'UPDATE';

const BASE_URL = 'http://localhost:3001';

//TODO more error handling around failed requests and bad response data

export default {
  [GET_LIST]: async ({ commit }) => {
    const response = await axios.get(`${BASE_URL}/list`);
    commit(UPDATE_LIST, response.data);
  },
  [ADD]: async ({ commit }, itemName) => {
    const result = await axios.post(`${BASE_URL}/add`, { title: itemName });
    commit(UPDATE_LIST, result.data);
  },
  [UPDATE]: async ({ commit }, item) => {
    const result = await axios.put(`${BASE_URL}/item/${item.id}`, item);
    commit(UPDATE_LIST, result.data);
  },
};
