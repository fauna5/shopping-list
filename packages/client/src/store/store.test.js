import createStore from './store';
import axios from 'axios';

import { ADD, GET_LIST, UPDATE } from './actions';

jest.mock('axios', () => ({
  get: jest.fn(),
  post: jest.fn(),
  put: jest.fn(),
}));

describe('store', () => {
  it('is set up correctly', () => {
    const store = createStore();
    expect(store.state).toMatchSnapshot();
  });

  it('is gets the list', async () => {
    const list = [
      {
        id: 0,
        title: 'eggs',
        complete: false,
      },
    ];
    axios.get.mockImplementation(() => ({
      data: list,
    }));

    const store = createStore();
    await store.dispatch(GET_LIST);
    expect(axios.get).toHaveBeenCalledWith('http://localhost:3001/list');
    expect(store.state.list).toEqual(list);
  });

  it('it adds to the list', async () => {
    const listAdded = [
      {
        id: 0,
        title: 'eggs',
        complete: false,
      },
      {
        id: 1,
        title: 'added',
        complete: false,
      },
    ];
    axios.post.mockImplementation((_, item) => ({
      data: listAdded,
    }));

    const store = createStore();
    await store.dispatch(ADD, 'new');
    expect(axios.post).toHaveBeenCalledWith('http://localhost:3001/add', {
      title: 'new',
    });
    expect(store.state.list).toEqual(listAdded);
  });

  it('it updates the status of the list items', async () => {
    const listUpdated = [
      {
        id: 0,
        title: 'eggs',
        complete: true,
      },
    ];
    axios.put.mockImplementation((_, item) => ({
      data: listUpdated,
    }));

    const store = createStore();
    await store.dispatch(UPDATE, { id: 0, complete: true });
    expect(axios.put).toHaveBeenCalledWith('http://localhost:3001/item/0', {
      id: 0,
      complete: true,
    });
    expect(store.state.list).toEqual(listUpdated);
  });
});
