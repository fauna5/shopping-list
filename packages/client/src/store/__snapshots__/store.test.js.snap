// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store is set up correctly 1`] = `
Object {
  "list": Array [],
}
`;
